src/core/transport/chttp2/incoming_metadata.c

```diff
@@ -0,0 +1,207 @@
+    grpc_chttp2_incoming_metadata_buffer *buffer, grpc_stream_op_buffer *sopb,
+    grpc_chttp2_incoming_metadata_live_op_buffer *live_op_buffer) {
+  grpc_stream_op *ops = sopb->ops;
+  size_t nops = sopb->nops;
+  size_t i;
+  size_t j;
+  size_t mdidx = 0;
+  size_t last_mdidx;
+  int found_metadata = 0;
+
+  /* rework the array of metadata into a linked list, making use
+     of the breadcrumbs we left in metadata batches during
+     add_metadata_batch */
+  for (i = 0; i < nops; i++) {
+    grpc_stream_op *op = &ops[i];
+    if (op->type != GRPC_OP_METADATA) continue;
+    found_metadata = 1;
+    /* we left a breadcrumb indicating where the end of this list is,
+       and since we add sequentially, we know from the end of the last
+       segment where this segment begins */
+    last_mdidx = (size_t)(gpr_intptr)(op->data.metadata.list.tail);
+    GPR_ASSERT(last_mdidx > mdidx);
+    GPR_ASSERT(last_mdidx <= buffer->count);
+    /* turn the array into a doubly linked list */
+    op->data.metadata.list.head = &buffer->elems[mdidx];
+    op->data.metadata.list.tail = &buffer->elems[last_mdidx - 1];
+    for (j = mdidx + 1; j < last_mdidx; j++) {
+      buffer->elems[j].prev = &buffer->elems[j - 1];
+      buffer->elems[j - 1].next = &buffer->elems[j];
+    }
+    buffer->elems[mdidx].prev = NULL;
+    buffer->elems[last_mdidx - 1].next = NULL;
+    /* track where we're up to */
+    mdidx = last_mdidx;
+  }
+  if (found_metadata) {
+    live_op_buffer->elems = buffer->elems;
+    if (mdidx != buffer->count) {
+      /* we have a partially read metadata batch still in incoming_metadata */
+      size_t new_count = buffer->count - mdidx;
+      size_t copy_bytes = sizeof(*buffer->elems) * new_count;
+      GPR_ASSERT(mdidx < buffer->count);
+      buffer->elems = gpr_malloc(copy_bytes);
+      memcpy(live_op_buffer->elems + mdidx, buffer->elems, copy_bytes);
+      buffer->count = buffer->capacity = new_count;
+    } else {
+      buffer->elems = NULL;
+      buffer->count = 0;
+      buffer->capacity = 0;
+    }
+  }
+}
+
+#if 0
+void grpc_chttp2_parsing_add_metadata_batch(
+    grpc_chttp2_transport_parsing *transport_parsing,
+    grpc_chttp2_stream_parsing *stream_parsing) {
+  grpc_metadata_batch b;
+
+  b.list.head = NULL;
+  /* Store away the last element of the list, so that in patch_metadata_ops
+     we can reconstitute the list.
+     We can't do list building here as later incoming metadata may reallocate
+     the underlying array. */
+  b.list.tail = (void *)(gpr_intptr)stream_parsing->incoming_metadata_count;
+  b.garbage.head = b.garbage.tail = NULL;
+  b.deadline = stream_parsing->incoming_deadline;
+  stream_parsing->incoming_deadline = gpr_inf_future;
+
+  grpc_sopb_add_metadata(&stream_parsing->data_parser.incoming_sopb, b);
+}
+#endif
+
+#if 0
+static void patch_metadata_ops(grpc_chttp2_stream_global *stream_global,
+                               grpc_chttp2_stream_parsing *stream_parsing) {
+}
+#endif
```

Comment:
Would you like me to review the "#if 0" code, or would you rather just delete it? If it is intended to be used later, I can review it, but if it is a vestige, please delete.